File: lim/전처리/전처리 - 균일하게 나누기 2.py
'''
행렬 누적합으로 풀이함
'''
import sys


def sol(n, grid):
    MAX_POS = 1000
    for i in range(1, MAX_POS + 1):
        for j in range(1, MAX_POS + 1):
            grid[i][j] += grid[i][j-1]
        for j in range(1, MAX_POS + 1):
            grid[i][j] += grid[i-1][j]

    def get_max_m(a, b):
        quarters = [0] * 4
        quarters[0] = grid[a][MAX_POS] - grid[a][b]
        quarters[1] = grid[a][b]
        quarters[2] = grid[MAX_POS][b] - grid[a][b]
        quarters[3] = (
            grid[MAX_POS][MAX_POS]
            - grid[a][MAX_POS]
            - grid[MAX_POS][b]
            + grid[a][b]
        )
        return max(quarters)

    min_max_m = sys.maxsize
    for i in range(2, MAX_POS + 1, 2):
        for j in range(2, MAX_POS + 1, 2):
            min_max_m = min(min_max_m, get_max_m(i, j))
    print(min_max_m)


n = int(sys.stdin.readline())
grid = [[0] * 1_001 for _ in range(1_001)]
points = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
for x, y in points:
    grid[x][y] = 1

sol(n, grid)

File: lim/전처리/전처리_균일하게_나누기2.java
package lim.전처리;

import java.util.*;
import java.io.*;

public class 전처리_균일하게_나누기2 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n;
    static final int MAX_VALUE = 1000;
    static int[][] grid, yPresumByX;

    static void sol() throws IOException {
        for (int i = 1; i < grid.length; i++) {
            for (int j = 1; j < grid[0].length; j++) {
                yPresumByX[i][j] = yPresumByX[i][j - 1] + grid[i][j];
            }
        }

        int minM = Integer.MAX_VALUE;
        for (int j = 2; j < grid[0].length; j += 2) {
            int downL = 0, downR = 0;
            for (int i = 2; i < grid.length; i += 2) {
                downL += yPresumByX[i - 1][j];
                downR += yPresumByX[i - 1][MAX_VALUE] - yPresumByX[i - 1][j];
            }

            int upL = 0, upR = 0;
            for (int i = 2; i < grid.length; i += 2) {
                downL -= yPresumByX[i - 1][j];
                downR -= yPresumByX[i - 1][MAX_VALUE] - yPresumByX[i - 1][j];
                upL += yPresumByX[i - 1][j];
                upR += yPresumByX[i - 1][MAX_VALUE] - yPresumByX[i - 1][j];

                int m = Math.max(upL, upR);
                m = Math.max(m, downR);
                m = Math.max(m, downL);

                minM = Math.min(minM, m);
            }
        }
        bw.write(minM + "");
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());

        grid = new int[MAX_VALUE + 1][MAX_VALUE + 1];
        yPresumByX = new int[MAX_VALUE + 1][MAX_VALUE + 1];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            grid[x][y]++;
        }
        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}
